src/server.c

```diff
@@ -4509,13 +4510,20 @@ sds genRedisInfoString(const char *section) {
         }
 
         if (server.loading) {
-            double perc;
+            double perc = 0;
             time_t eta, elapsed;
-            off_t remaining_bytes = server.loading_total_bytes-
-                                    server.loading_loaded_bytes;
-
-            perc = ((double)server.loading_loaded_bytes /
-                   (server.loading_total_bytes+1)) * 100;
+            off_t remaining_bytes = 1;
+
+            if (server.loading_total_bytes) {
+                perc = ((double)server.loading_loaded_bytes / server.loading_total_bytes) * 100;
+                remaining_bytes = server.loading_total_bytes - server.loading_loaded_bytes;
+            } else if(server.loading_rdb_used_mem) {
+                perc = ((double)server.loading_loaded_bytes / server.loading_rdb_used_mem) * 100;
```

Comment:
how about `zmalloc_used_memory()` instead of `server.loading_loaded_bytes`

Reply:
Please ignore, `zmalloc_used_memory` will include last db size if `repl-diskless-load` type is swapdb

Reply:
My logic is that rdb is roughly the same size of the used memory. (one is encoded more efficiently, but has some encoding overheads, the other has structs and pointers) 

But I like your idea. 
for swapdb, maybe we can make a snapshot of  the used memory at the beginning of loading, and use the delta from the current?

Reply:
But we also need to consider `lazy flush`, that brings much complexity

Reply:
forgot about that, that's really nearly impossible.
only thing we can do, which is ugly, is use used_memory in some cases, and loaded_bytes in other cases.
it'll make the estimated progress more accurate on many cases, but code would be ugly

Reply:
i'd like another opinion. @guybe7 ?

Reply:
i feel that in this case we should compromise on the accuracy in order to keep the code simple.. i.e. use `loading_loaded_bytes`

Reply:
I agree with @guybe7. I would even argue that we shouldn't present heuristics as metrics - just tell what we know (i.e. `loading_rdb_used_mem`) and leave it to users to apply heuristics.

Reply:
@yossigo if i understand Guy's comment he meant that he's ok with the current code and the inaccurate heuristics (keeping the code simple).

the way i see it the `loading_loaded_perc` and `loading_eta_seconds` are not real metrics that can be trusted, but rather just a nice thing for humans to look at when using redis-cli.

also, if someone needs these heuristics (again for human eyes) i think it's slightly better to have them in redis rather than for example document this (a bit complicated formula) so that someone can do that outside (handling both disk-based and diskless in an inaccurate / best effort way).

Reply:
@oranagra That's exactly what I meant. I'm OK with less accurate heuristics but simpler code. I'm even MORE OK with no heuristics, although I agree with your point of human users. The problem here is to properly communicate what is a hard metric (to be consumed by monitoring systems) and what is not. We could do that through explicit naming but that's water under the bridge now...